torch_geometric/loader/link_loader.py

```diff
@@ -0,0 +1,217 @@
+            **kwargs,
+        )
+
+    def filter_fn(
+        self,
+        out: Union[SamplerOutput, HeteroSamplerOutput],
+    ) -> Union[Data, HeteroData]:
+        r"""Joins the sampled nodes with their corresponding features,
+        returning the resulting (Data or HeteroData) object to be used
+        downstream."""
+        if isinstance(out, SamplerOutput):
+            edge_label_index, edge_label = out.metadata
+            data = filter_data(self.data, out.node, out.row, out.col, out.edge,
+                               self.link_sampler.edge_permutation)
+            data.edge_label_index = edge_label_index
+            data.edge_label = edge_label
+
+        elif isinstance(out, HeteroSamplerOutput):
+            edge_label_index, edge_label, edge_label_time = out.metadata
+            if isinstance(self.data, HeteroData):
+                data = filter_hetero_data(self.data, out.node, out.row,
+                                          out.col, out.edge,
+                                          self.link_sampler.edge_permutation)
+            else:  # Tuple[FeatureStore, GraphStore]
+                data = filter_custom_store(*self.data, out.node, out.row,
+                                           out.col, out.edge)
+
+            edge_type = self.input_type
+            data[edge_type].edge_label_index = edge_label_index
+            data[edge_type].edge_label = edge_label
+            if edge_label_time is not None:
+                data[edge_type].edge_label_time = edge_label_time
+
+        else:
+            raise TypeError(f"'{self.__class__.__name__}'' found invalid "
+                            f"type: '{type(out)}'")
+
+        return data if self.transform is None else self.transform(data)
+
+    def collate_fn(self, index: EdgeSamplerInput) -> Any:
+        r"""Samples a subgraph from a batch of input nodes."""
+        out = self.link_sampler.sample_from_edges(
+            index,
+            negative_sampling_ratio=self.neg_sampling_ratio,
+        )
+        if self.filter_per_worker:
+            # We execute `filter_fn` in the worker process.
+            out = self.filter_fn(out)
+        return out
+
+    def _get_iterator(self) -> Iterator:
+        if self.filter_per_worker:
+            return super()._get_iterator()
+        # We execute `filter_fn` in the main process.
+        return DataLoaderIterator(super()._get_iterator(), self.filter_fn)
+
+    def __repr__(self) -> str:
+        return f'{self.__class__.__name__}()'
+
+
+###############################################################################
+
+
+class Dataset(torch.utils.data.Dataset):
+    def __init__(self, edge_label_index: torch.Tensor,
+                 edge_label: torch.Tensor, edge_label_time: OptTensor = None):
+        self.edge_label_index = edge_label_index
+        self.edge_label = edge_label
+        self.edge_label_time = edge_label_time
+
+    def __getitem__(self, idx: int) -> Tuple[int]:
```

Comment:
Looks like we actually return `Union[Tuple[Tensor, Tensor, Tensor], Tuple[Tensor, Tensor, Tensor, Tensor]]`.